### DISCUSSION TITLE
Two binding sets bound with different types for key

### DISCUSSION THREAD

**Discussion by Wdestroier** [activej/activej#160] [Q&A] [answered]
Hi! I'm trying to specify a custom server port, but an exception is being thrown. Any idea what may be the reason why? I grabbed this example from https://activej.io/http/examples#websocket-pong-server.

`Exception in thread "main" io.activej.inject.binding.DIException: Two binding sets bound with different types for key AsyncHttpServer`

Code to reproduce the problem:
```java
public class ExampleApplication extends HttpServerLauncher {

    public static void main(String... args) throws Exception {
        var app = new ExampleApplication();
        app.launch(args);
    }

    @Provides
    public AsyncServlet rootServlet() {
        return httpRequest -> HttpResponse.ok200();
    }

    @Eager
    @Provides
    public AsyncHttpServer server(Eventloop eventloop, AsyncServlet rootServlet, Config config) {
        return AsyncHttpServer.create(eventloop, rootServlet)
                .withInitializer(ofHttpServer(config.getChild("http")))
                ;//.withListenPort();
    }

}
```

It would be nice if the server port could come from a config file (`example.properties`). Does the Config need to be loaded twice, in the server() method and in the getBusinessLogicModule() method, or can it be loaded just once?

The rootServlet is not needed. All servlets are specified in the modules that are loaded by overriding `protected Module getBusinessLogicModule() {}`. If I remove this servlet, then another exception is thrown.
```
Exception in thread "main" io.activej.inject.binding.DIException: Unsatisfied dependencies detected:
	key AsyncServlet required to make:
		- AsyncHttpServer at object ExampleApplication@58c1c010, provider method HttpServerLauncher.server(HttpServerLauncher.java:0)
```

Is it possible to remove the rootServlet method that is providing a dummy endpoint?

**Comment by eduard-vasinskyi** [comment, marked as answer]
Hi, @Wdestroier

You get the `Two binding sets bound with different types for key…` exception because you have conflicting bindings for the `AsyncHttpServer` key. There is a defined binding in `HttpServerLauncher#server` method and you have defined another binding in `ExampleApplication#server` method. If you would like to override the default binding from `HttpServerLauncher` you should’ve overridden the `Launcher#getOverrideModule` method and provide your binding inside the module returned from the `Launcher#getOverrideModule` method.

But if all you want to do is just specify a different port value you can use the `Config` class and specify a different value for the `http.listenAddresses` property. You could do this in 2 ways: 
- Use the `http-server.properties` file and specify a listening address there
- Pass the VM option `-Dconfig.http.listenAddresses=<your listen address>`

[HttpServerLauncher](https://github.com/activej/activej/blob/master/launchers/http/src/main/java/io/activej/launchers/http/HttpServerLauncher.java) is an example launcher that you can use as a basis for simple HTTP server launchers. For more complex launchers we recommend to copy-paste the code from the `HttpServerLauncher` and use that as a starting point for building your own custom Launcher.

As for the `Unsatisfied dependencies detected` exception, it is thrown because there is no provided binding for the `AsyncServlet`. If you have specified multiple servlets in different modules, make sure that one of these servlets is bound as `AsyncServlet`. A common use case is to have multiple servlets provided where each servlet is mapped to a different route using [RoutingServlet](https://github.com/activej/activej/blob/master/core-http/src/main/java/io/activej/http/RoutingServlet.java). A `RoutingServlet` is itself bound as an `AsyncServlet` so that it can be picked up by the `AsyncHttpServer` binding. (see the [RoutingServlet example](https://activej.io/http/examples#routing-servlet-example) for reference).

Please, let me know if you have any more questions about ActiveJ.

**Comment by Wdestroier** [comment]
> HttpServerLauncher is an example launcher that you can use as a basis for simple HTTP server launchers

Got it, thanks!

> make sure that one of these servlets is bound as AsyncServlet

I'm providing an AsyncServlet in the controller, but for some reason the servlet cannot be found... Are the modules bound incorrectly?

```java
public class ExampleApplication extends Launcher {
    // ...

    @Override
    protected final Module getModule() {
        return combine(
                ServiceGraphModule.create(),
                ConfigModule.create().withEffectiveConfigLogger(),
                getExampleModule()
        );
    }

    protected Module getExampleModule() {
        return ModuleBuilder.create()
                    .bind(ObjectMapper.class).to(ObjectMapper::new)
                    .bind(ExampleController.class)
                    .build();
    }
}

@Inject
public class ExampleController extends AbstractModule {
    @Inject
    private ObjectMapper objectMapper;

    @Provides
    public AsyncServlet servlet() {
        return RoutingServlet.create()
                .map(GET, "/example", request -> HttpResponse.ok200()
                        .withJson(objectMapper.writeValueAsString("{}")));
    }
}
```

**Comment by eduard-vasinskyi** [comment]
Since `ExampleController` is a [Module](https://github.com/activej/activej/blob/master/core-inject/src/main/java/io/activej/inject/module/Module.java) (extends [AbstractModule](https://github.com/activej/activej/blob/master/core-inject/src/main/java/io/activej/inject/module/AbstractModule.java)) you need to instantiate it and pass the `ExampleController` instance to the `getModule()` method somehow. You can simply add it to the `combine(Module...)` method like this:
```java
@Override
protected final Module getModule() {
     return combine(
           ServiceGraphModule.create(),
           ConfigModule.create().withEffectiveConfigLogger(),
           getExampleModule(),
           new ExampleController()
     );
}
```
Or alternatively you can install it inside the `getExampleModule()` method:
```java
protected Module getExampleModule() {
  return ModuleBuilder.create()
        .bind(ObjectMapper.class).to(ObjectMapper::new)
        .install(new ExampleController())
        .build();
}
```
(Notice that I removed the `.bind(ExampleController.class)` piece, I’ll explain it later).

You have marked the `ExampleController` class as well as its `objectMapper` field with `@Inject` annotation. But ActiveJ Inject only injects fields into instances that were created by the [Injector](https://github.com/activej/activej/blob/master/core-inject/src/main/java/io/activej/inject/Injector.java). And we have instantiated `ExampleController` manually (see above snippets) as Injector needs to know about this module upon instantiation.

What you can do instead is remove `@Inject` annotations from the `ExampleController` class and add `ObjectMapper` as a parameter of the `ExampleController#servlet` method:
```java
public class ExampleController extends AbstractModule {
  @Provides
  public AsyncServlet servlet(ObjectMapper objectMapper) {
     return RoutingServlet.create()
           .map(GET, "/example", request -> HttpResponse.ok200()
                 .withJson(objectMapper.writeValueAsString("{}")));
  }
}
```
If you rewrite `ExampleController` class like this you would not need to add the `.bind(ExampleController.class)` piece.

_PS. It is uncommon to have `@Inject`-annotated fields inside an `AbstractModule`. However, if you absolutely must do this, you could post-inject your `ExampleController` instance after it has been passed to the `getModule()` method. To do this you would have to use [InstanceInjector\<ExampleController>](https://github.com/activej/activej/blob/master/core-inject/src/main/java/io/activej/inject/InstanceInjector.java) and inject the instance manually. See the [example](https://activej.io/inject/injecting-instances#using-an-instanceinjector)._

**Comment by Wdestroier** [comment]
> It is uncommon to have @Inject-annotated fields inside an `AbstractModule`.

These classes extend AbstractModule to make them injectable...

Is it possible to bind a type to another? Example: bind ExampleService to ExampleServiceImpl.
Or should I bind ExampleServiceImpl directly?

```java
@Inject
@NoArgsConstructor
public class ExampleController extends AbstractModule { // This class extends AbstractModule to load the route only
    @Inject
    private ObjectMapper objectMapper;
    @Inject
    private ExampleService exampleService; // This service isn't injected

    @Provides
    public AsyncServlet servlet() {
        return RoutingServlet.create().map(GET, "/", request -> HttpResponse.ok200().withJson("{}"));
    }
}

@Inject
@NoArgsConstructor
public class ExampleServiceImpl extends AbstractModule implements ExampleService {
    @Inject
    private Config config;
}

protected Module getBusinessLogicModule() {
    return ModuleBuilder.create()
        .bind(ObjectMapper.class).to(ObjectMapper::new)
        .install(new ExampleController())
        // .bind(ExampleService.class).to(ExampleServiceImpl::new) // Doesn't inject the dependencies in ExampleServiceImpl
        // .install(new ExampleServiceImpl()) // Doesn't inject the ExampleServiceImpl in ExampleController
        .build();
}
```

**Comment by eduard-vasinskyi** [comment]
Modules are not intended to be provided in runtime or injected into runtime components. (Modules can be viewed as pieces of start-up configuration for compiling DI graph, prior to any runtime injection taking place). Therefore it does not make any practical sense to annotate them with `@Inject` annotation and make them injectable.
 
Also please note, DI injections are taking place only into the components created/provided/managed by DI. But if you create any class instance by yourself with a `new` Java keyword, it’s just plain Java - that’s why nothing is injected into those fields.

Please do not try to inject anything into modules. 

Look at our extensive [examples for instance injecting](https://activej.io/inject/injecting-instances).

Injected classes do not extend `AbstractModule`, neither should your injected classes. You need to extend `AbstractModule` if you want to add bindings using `@Provides` annotated methods.

In fact we may even forbid `@Inject` annotations on DI modules in future releases of ActiveJ, to avoid confusion. 

> Is it possible to bind a type to another? Example: bind ExampleService to ExampleServiceImpl.

Yes! You absolutely can bind interfaces to the implementation. But rather than binding to a constructor (`ExampleServiceImpl::new`) just bind it to a class (and remove `extends AbstractModule` from the `ExampleServiceImpl`). This way the Injector will inject a Config instance into ExampleServiceImpl. Here is how a binding will look like:
```java
.bind(ExampleService.class).to(ExampleServiceImpl.class)
```

_PS. I forgot to add `@Provides` annotation to the `ExampleController#servlet` method in a previous reply._

**Comment by Wdestroier** [comment]
Thanks, Edward! 😃
